### PR TITLE
Zth cipher class

### PR DESCRIPTION
**Pull request by zachholcomb**
Add cipher class
Add enigma child of cipher class
Redistribute responsibility of methods to follow single responsibility

### CODE REVIEW

File: lib/cipher.rb [hunk 1]
```diff
@@ -1,9 +1,66 @@
...
+
   def initialize
-    @shift_key = ShiftGenerator.new
+    @default_key = generate_random_key
+    @default_date = create_date
+    @shift_generator = ShiftGenerator.new
+    @character_set = create_character_set
+  end
+
+  def chop_message(message)
```
Comment (zachholcomb) [inline lib/cipher.rb:19]: chop message splits message to encode or decode into usable arrays

File: lib/cipher.rb [hunk 2]
```diff
@@ -1,9 +1,66 @@
...
+    message.downcase.split("").each_slice(4).to_a
+  end
+
+  def encrypt_letter(letter, shift)
+    encrypted_set = Hash[@character_set.zip(@character_set.rotate(shift))]
```
Comment (zachholcomb) [inline lib/cipher.rb:24]: this method encrypts one letter at a time

File: lib/cipher.rb [hunk 3]
```diff
@@ -1,9 +1,66 @@
...
+    return letter if !encrypted_set.keys.include?(letter)
+    encrypted_set.fetch(letter)
+  end
+
+  def encrypt_chop(chop, shift_key)
+    chop.map.with_index do |letter, index|
```
Comment (zachholcomb) [inline lib/cipher.rb:30]: this method encrypts the one array of the message at a time, calling on encrypt letter with the correct shift value

File: lib/cipher.rb [hunk 4]
```diff
@@ -1,9 +1,66 @@
...
+      encrypt_letter(letter, shift_key[index])
+    end
+  end
+
+  def encrypt_message(message, shift)
+    chop_message(message).map do |chunk|
```
Comment (zachholcomb) [inline lib/cipher.rb:36]: this method puts the above two together. the reasoning for this is to keep method lengths shorter and in granular chunks that makes sense to me, might be a tad redundant though, will consider refactoring

File: lib/cipher.rb [hunk 5]
```diff
@@ -1,9 +1,66 @@
...
+      encrypt_chop(chunk, shift)
+    end.join
+  end
+
+  def decrypt_letter(letter, shift)
```
Comment (zachholcomb) [inline lib/cipher.rb:41]: decrypt methods are same as encrypt methods but use a negative shift value

File: lib/cipher.rb [hunk 6]
```diff
@@ -1,9 +1,66 @@
...
+    end
+  end
+
+  def decrypt_message(message, shift)
+    chop_message(message).map do |chunk|
+      decrypt_chop(chunk, shift)
+    end.join
+  end
+
+  def input_verification(key = self.default_key, date = self.default_date)
```
Comment (zachholcomb) [inline lib/cipher.rb:59]: input verification can take a key or a date and will validate if key or date are a key or a date and calls on shift generator to create the shift key

File: lib/cipher.rb [hunk 7]
```diff
@@ -1,9 +1,66 @@
...
+    if !@shift_generator.key_gen.verify_key(key)
+      date = key
+      key = self.default_key
+    end
+    [@shift_generator.shift_key(key, date), key, date]
```
Comment (zachholcomb) [inline lib/cipher.rb:64]: the output of this method is an array so that enigma's encrypt method can access the correct values

File: lib/enigma.rb [hunk 8]
```diff
@@ -1,8 +1,24 @@
-class Enigma
-  attr_reader :shift_generator, :character_set
+require_relative 'cipher'
+
+class Enigma < Cipher
```
Comment (zachholcomb) [inline lib/enigma.rb:3]: enigma is a child of cipher, implemented so that future encrypters could theoretically exist.  for example a cesar cipher could also be a child of cipher

File: lib/enigma.rb [hunk 9]
```diff
@@ -1,8 +1,24 @@
...
-    @shift_generator = ShiftGenerator.new
+    super
+  end
+
+  def encrypt(message, key = self.default_key, date = self.default_date)
+    shift = input_verification(key, date)
+    { encryption: encrypt_message(message, shift[0]),
+      key: shift[1],
+      date: shift[2]
+    }
```
Comment (zachholcomb) [inline lib/enigma.rb:14]: encrypt method calls on methods defined in cipher so that the sole responsibility of encrypt is to take input messages and output messages along with the keys used to encrypt the messages

File: lib/key_gen.rb [hunk 10]
```diff
@@ -3,17 +3,16 @@
...
+    ]
   end
 
-  def create_key(five_digit_key)
-    { key_A: five_digit_key[0..1].to_i,
-      key_B: five_digit_key[1..2].to_i,
-      key_C: five_digit_key[2..3].to_i,
-      key_D: five_digit_key[3..4].to_i
-    }
+  def verify_key(key)
```
Comment (zachholcomb) [inline lib/key_gen.rb:15]: added a verify key query method to check if the key is the right length

File: lib/modules/create_defaults.rb [hunk 11]
```diff
@@ -1,9 +1,17 @@
 module Generable
+  def create_character_set
+    ("a".."z").to_a << " "
+  end
```
Comment (zachholcomb) [inline lib/modules/create_defaults.rb:4]: removed character set from cipher to generable modules

File: lib/offset_gen.rb [hunk 12]
```diff
@@ -16,10 +8,10 @@ def last_four_digits(squared_date_param)
...
-    { key_A: last_four_digits_param[0].to_i,
-      key_B: last_four_digits_param[1].to_i,
-      key_C: last_four_digits_param[2].to_i,
-      key_D: last_four_digits_param[3].to_i
-    }
+    [ last_four_digits_param[0].to_i,
+      last_four_digits_param[1].to_i,
+      last_four_digits_param[2].to_i,
+      last_four_digits_param[3].to_i
+    ]
```
Comment (zachholcomb) [inline lib/offset_gen.rb:15]: was originally using a hash here, but an array was the better choice to implement encryption

File: lib/shift_generator.rb [hunk 13]
```diff
@@ -2,18 +2,19 @@
...
 
 class ShiftGenerator
-  attr_reader :key, :offset
+  attr_reader :key_gen, :offset_gen, :shift
 
   def initialize
-    @key = KeyGen.new
-    @offset = OffsetGen.new
+    @key_gen = KeyGen.new
+    @offset_gen = OffsetGen.new
```
Comment (zachholcomb) [inline lib/shift_generator.rb:9]: changed variable names here so that they could be distinguished between a key param and a key generator

File: lib/shift_generator.rb [hunk 14]
```diff
@@ -2,18 +2,19 @@
...
-    random_key_hash = @key.create_key(five_digit_key)
-    date_offset_hash = @offset.create_offset(last_four_digits_param)
-    random_key_hash.merge(date_offset_hash) do |key, random_value, date_value|
-      random_value + date_value
+  def shift_key(five_digit_key, date)
+    key_param = @key_gen.create_key(five_digit_key)
+    date_param = @offset_gen.last_four_digits(@offset_gen.square_date(date))
+
+    key_param.zip(@offset_gen.create_offset(date_param)).map do |pair|
+      pair.sum
```
Comment (zachholcomb) [inline lib/shift_generator.rb:17]: this output was originally a hash but changed to an array to implement functionality

File: test/enigma_test.rb [hunk 15]
```diff
@@ -18,46 +16,124 @@ def test_it_has_attributes
...
+    @enigma.stubs(:default_date).returns("040895")
+    @enigma.stubs(:default_key).returns("02715")
+    expected = {
+      decryption: "hello world!",
+      key: "02715",
+      date: "040895"
+    }
+    assert_equal expected, @enigma.decrypt("keder ohulw!")
+  end
+end
```
Comment (zachholcomb) [inline test/enigma_test.rb:139]: tried to test as many edge cases as I could think of, one issue is that not sure if special characters should take a shift index spot.  Not sure how to implement that functionality though, will try to add a guard clause
